kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoWindowedTable.java

```diff
@@ -0,0 +1,489 @@
+
+    LOG.info("{}[{}] Updating offset to {} and streamTime to {} with epoch {}",
+             name, kafkaPartition, offset, streamTime, epoch);
+
+    return new UpdateOneModel<>(
+        Filters.and(
+            Filters.eq(WindowMetadataDoc.ID, kafkaPartition),
+            Filters.lte(WindowMetadataDoc.EPOCH, epoch)
+        ),
+        Updates.combine(
+            Updates.set(WindowMetadataDoc.OFFSET, offset),
+            Updates.set(WindowMetadataDoc.STREAM_TIME, streamTime),
+            Updates.set(WindowMetadataDoc.EPOCH, epoch)
+        )
+    );
+  }
+
+  public long epoch(final int kafkaPartition) {
+    return kafkaPartitionToSegments.get(kafkaPartition).epoch;
+  }
+
+  /**
+   * Inserts data into this table. Note that this will overwrite
+   * any existing entry in the table with the same key.
+   *
+   * @param kafkaPartition the kafka partition
+   * @param windowedKey    the windowed data key
+   * @param value          the data value
+   * @param epochMillis    the timestamp of the event
+   * @return a statement that, when executed, will insert the value
+   *         for this key and partition into the table.
+   *         Note that the key in this case is a "windowed" key, where
+   *         {@code windowedKey} includes both the record key and
+   *         the windowStart timestamp
+   */
+  @Override
+  public WriteModel<WindowDoc> insert(
+      final int kafkaPartition,
+      final WindowedKey windowedKey,
+      final byte[] value,
+      final long epochMillis
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+    partitionSegments.segmentBatch.updateStreamTime(windowedKey.windowStartMs);
+
+    final long epoch = partitionSegments.epoch;
+    return new UpdateOneModel<>(
+        Filters.and(
+            Filters.eq(WindowDoc.ID, windowedKey.key.get()),
+            Filters.eq(WindowDoc.WINDOW_START_TS, windowedKey.windowStartMs),
+            Filters.lte(WindowDoc.EPOCH, epoch)
+        ),
+        Updates.combine(
+            Updates.set(WindowDoc.VALUE, value),
+            Updates.set(WindowDoc.EPOCH, epoch)
+        ),
+        UPSERT_OPTIONS
+    );
+  }
+
+  /**
+   * @param kafkaPartition  the kafka partition
+   * @param windowedKey    the windowed data key
+   *
+   * @return a statement that, when executed, will delete the value
+   *         for this key and partition in the table.
+   *         Note that the key in this case is a "windowed" key, where
+   *         {@code windowedKey} includes both the record key and
+   *         the windowStart timestamp
+   */
+  @Override
+  public WriteModel<WindowDoc> delete(
+      final int kafkaPartition,
+      final WindowedKey windowedKey
+  ) {
+    final var partitionSegments = kafkaPartitionToSegments.get(kafkaPartition);
+    partitionSegments.segmentBatch.updateStreamTime(windowedKey.windowStartMs);
+    throw new UnsupportedOperationException("Deletes not yet supported for MongoDB window stores");
```

Comment:
is this necessary to implement for the DSL operators?

Reply:
Technically it's possible to invoke this since our store implementation delegates to `delete` when it receives an insert with a null value. It's probably rare but still very much possible (and definitely allowed) for an aggregation to return `null` so I definitely intend to implement this at some point. However based on the code snippet of the aggregator app, they definitely don't/can't return `null`, so I decided to put this off for the first iteration of the window store.

tl;dr yes we should implement it, no we don't need to ASAP